### ISSUE TITLE
Data Volume grown when ran Compaction API 

### ISSUE THREAD

**Issue by sanjiv1980** [delta-io/delta#254]
I observed like when I ran the compaction API , my data volume has grown up to 5-times more than that . I just wanted to know It is happening because of more metadata are getting added into file .?

**Comment by tdas** [comment]
how are you measuring the data volume? sum total of all the files in the directory?

**Comment by sanjiv1980** [comment]
yes you right , before doing compaction , i checked the volume and after compaction I checked the volume . Could you please explain like why it grown-up .
Thanks,

**Comment by sanjiv1980** [comment]
I followed below step to reproduce 
formed a delta lake from input of size 121 MB , It became 128MB.
Did some merge operation after that It became 180 MB (got so many small files).
Ran Vacuum API to delete those removed file , after that It became 157 MB.
Ran Compaction API, after that It became 478MB.
Ran Vacuum API to delete those removed file , after that It became 322 MB.

**Comment by tdas** [comment]
Delta using multi-version concurrency control, so it maintains data files for previous versions of the table. So when you are compacting, it writing new data files to create a new version of the table, but it is still keeping around the previous version's data files. Hence, just be directory listing, it will look like the size multiplied. But in reality, only some of the files are used by the latest version of the table, and in case of compaction, those latest version files should be the larger files. 

To see the real data size of only the latest version, you should vacuum with 0 retention (with caution, see https://docs.delta.io/0.4.0/delta-utility.html#vacuum). So in your test, did you run vacuum with 0 retention?

**Comment by sanjiv1980** [comment]
Yes , I did it , if you see the above my steps which I followed ( step 3 and step 5), but even-though the size is more than 3-times.
Please explain me , if possible .

**Comment by tdas** [comment]
That is indeed pretty odd. How did you do the compaction? Can you share the code? 
Also can you do something like a count before and after the compaction to double check that the compaction is correct?

**Comment by sanjiv1980** [comment]
```
SparkConf conf = new SparkConf().setMaster("local[*]").setAppName("SparkFileSumApp");
JavaSparkContext sc = new JavaSparkContext(conf);
SparkSession spark = SparkSession.builder().config(conf).getOrCreate();
DataFrameReader dataFrameReader = spark.read();
Dataset<Row> pData = dataFrameReader.parquet("/Users/kumars3/Documents/workspace-sts/espn-deltalake/input-data/3-sep/");
pData.write().mode("append").option("mergeSchema", "true").format("delta").save("delta-table/clickstream/");
```
1. Took Sample of 166MB of input data. : Total number Records : 1630168
2. Once Delta Lake formed It became of 176MB : Total number of Records : 1630168

**Comment by tdas** [comment]
This does not make sense to me. 
> Ran Compaction Now size became 781MB : Total number of records : 3260336 (double of original)
Your compaction code seems correct to me but how can it double the number of rows if you are overwriting the table. The only way that can happen is if instead of overwrite, you ran it in append mode, which would precisely duplicate the data.

**Comment by sanjiv1980** [comment]
Thats what I am doing . Please suggest , what can I do to figure it out . 
One thing I noted like , if I keep the `repartition` count ( less than 10) , then size is getting increase comparatively if you keep  repartition count bigger . Could you explain the logic behind this .

Thanks

**Comment by sanjiv1980** [comment]
@tdas Could you please help me to figure it out the issue .? I am using above code for all my activities , will be great help to fix my issue . Unable to fix it .

**Comment by Devian-ua** [comment]
@sanjiv1980 

- Try compression. e.g. snappy
- Try sorting instead of doing repartition. (I used sort by uid on my dataset, it saves 10%)

**Comment by andormarkus** [comment]
Hi @sanjiv1980 

After two years I might, can give you an answer. 

The documentation example uses `repartition` instead of `coalesce` to decrease the number of partitions -> number of files.
`repartition` does full data shuffle and your data will be quite unsorted in a partition. 

> In general columnar storage formats like Parquet are highly sensitive when it comes to data distribution (data organization) and cardinality of individual columns. The more organized is data and the lower is cardinality the more efficient is the storage.

In this [article](https://towardsdatascience.com/about-sort-in-spark-3-x-f3699cc31008) it is well defined how data sorting affect the final size of your file.

# close

Sources:
<https://stackoverflow.com/questions/54218006/why-does-the-repartition-method-increase-file-size-on-disk>
<https://stackoverflow.com/questions/38153935/why-are-spark-parquet-files-for-an-aggregate-larger-than-the-original>
<https://mrpowers.medium.com/managing-spark-partitions-with-coalesce-and-repartition-4050c57ad5c4#.36o8a7b5j>
<https://towardsdatascience.com/merging-too-many-small-files-into-fewer-large-files-using-apache-spark-in-datalake-ff9a32807056>
<https://towardsdatascience.com/about-sort-in-spark-3-x-f3699cc31008>
<https://towardsdatascience.com/notes-about-saving-data-with-spark-3-0-86ba85ca2b71>
<https://towardsdatascience.com/best-practices-for-bucketing-in-spark-sql-ea9f23f7dd53>
